{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylogenetics-script-Pseudomonas-Syringae_S.Pflanze"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Constructing phylogenetic trees for the unknown Pseudomonas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Author: AI \n",
    "\n",
    "Date: 01.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 00 - Conda env buildup and versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conda create env = Blasty\n",
    "conda create -y -n blasty -c conda-forge -c bioconda blast taxonkit biopython\n",
    "\n",
    "# Conda create env = phylogenetics\n",
    "conda create --name phylogenetics -c bioconda  gettools clustalo seqkit\n",
    "\n",
    "# Create conda env = bit\n",
    "conda create -n bit -c conda-forge -c bioconda -c defaults -c astrobiomike bit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01- Prepare Input Output and Scripts directories "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can find the github repository of the project [here](https://github.com/Darcy220606/pseudomonas_Spflanze.git)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "git clone https://github.com/Darcy220606/NRPS-Pnunensis4A2e.git"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Project directory**\n",
    "```\n",
    ".\n",
    "├── 01-resources\n",
    "├── 02-scripts\n",
    "├── 03-data\n",
    "├── 04-analysis\n",
    "├── 05-results\n",
    "├── 06-publication\n",
    "└── README.md\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Change into the directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /NRPS-Pnunensis4A2e"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 02- Pseudomonas refseq - Gtotree (complete and rep. genomes) \n",
    "#### Which strain is the closest? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is based on [gtotree](https://github.com/AstrobioMike/GToTree/wiki/example-usage#alteromonas-example)\n",
    "\n",
    "A. Find out where this strain belongs in the Pseudomonas refseq NCBI db by consructing a de novo phylogenomic tree.\n",
    "   \n",
    "   1. Download the root sequence manually. We chose [Aestuariirhabdus haliotis](https://www.ncbi.nlm.nih.gov/labs/data-hub/genome/GCF_023509475.1/) located in `04-analysis/gtotree/GCF_023509475.1_ASM2350947v1_genomic.fna`\n",
    "   \n",
    "   2. Download accession numbers as below. Located in `04-analysis/gtotree/assembly-accs-complete-genomes.txt` (1432)\n",
    "   \n",
    "   3. Create a list with genome_to_id_map by typing in the name of the root and the unknown genome : MANUALLY located in `/04-analysis/gtotree/genome_to_id_map.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "\n",
    "PROJECT_DIR=NRPS-Pnunensis4A2e\n",
    "REF_GENOME=/Net/Groups/ccdata/databases/ncbi-ref-genomes/pseudomonas/completegenomes #representative\n",
    "AN_DIR=$PROJECT_DIR/04-analysis/gtotree\n",
    "\n",
    "mkdir $REF_GENOME\n",
    "mkdir $AN_DIR\n",
    "\n",
    "##################################################################\n",
    "# NCBI: Download refgenomes of Pseudomonas :  27461(total) -to- 25510 (fasta) genomes = memory issues\n",
    "#                                             14,376 (total refseq) -to- 14377 (fasta) genomes = memory issues\n",
    "#                                             1432 (total refseq-comp genomes) -to- 1372 (fasta) genomes\n",
    "#                                             321 (total refseq-rep genomes) -to-  (fasta) genomes\n",
    "##################################################################\n",
    "conda activate blasty\n",
    "\n",
    "cd $REF_GENOME\n",
    "\n",
    "#OPTION#1: Grab the search query from the Assembly database on ncbi first  :: Pseudomonas-complete genomes-ref/gen\n",
    "esearch -db assembly -query '\"Pseudomonas\"[Organism] OR Pseudomonas[All Fields]) AND ((\"latest refseq\"[filter] OR \"latest genbank\"[filter]) AND \"complete genome\"[filter] AND all[filter] NOT anomalous[filter]' \\\n",
    "| esummary | xtract -pattern DocumentSummary -element AssemblyAccession > $AN_DIR/assembly-accs-complete-genomes.txt\n",
    "\n",
    "##################################################################\n",
    "# Create a list with genome_to_id_map\n",
    "##################################################################\n",
    "\n",
    "# Create a file with fasta and gbff locations with \n",
    "cd $AN_DIR\n",
    "ls *.fa > fasta_files.txt\n",
    "ls *.gbff > genbank_files.txt\n",
    "\n",
    "##################################################################\n",
    "# Construct the phylogenetic tree\n",
    "##################################################################\n",
    "\n",
    "conda activate gtotree\n",
    "\n",
    "cd $AN_DIR\n",
    "\n",
    "GToTree -a $AN_DIR/assembly-accs-complete-genomes.txt -g genbank_files.txt \\\n",
    "        -f fasta_files.txt -H Gammaproteobacteria \\\n",
    "        -t -L Species,Strain -m genome_to_id_map.tsv -j 11 \\\n",
    "        -o pseudomonas_refseq-compgenomes\n",
    "        \n",
    "conda deactivate\n",
    "\n",
    "# RUN results: \n",
    "# 1. 1 genome accession removed GCF_000902895.1 - for too few hits\n",
    "# 2. 1292 of the total 1372 input accessions had their genomes successfully downloaded and searched\n",
    "# 3. Visualise trees *.tre with itol\n",
    "\n",
    "##################################################################\n",
    "# Nucleutide based tree on the clade containing the 4A2E genome\n",
    "##################################################################\n",
    "# MANUALLY: prune the tree to the seq. on teh clade that contains the 4A2E genome (19 genomes)\n",
    "#Found here: `/04-analysis/gtotree/pruned_acc.txt`\n",
    "\n",
    "##################################################################\n",
    "# RUN fastANI and plot a tree  (ANI)\n",
    "##################################################################\n",
    "\n",
    "## Download the fasta files in the clade containing the genome isolate \n",
    "conda activate bit\n",
    "\n",
    "# Download the fasta files corresponnding to the accessions\n",
    "# -j parallelism \n",
    "ANI=$PROJECT_DIR/04-analysis/gtotree/ANI\n",
    "ACC=$PROJECT_DIR/04-analysis/gtotree\n",
    "\n",
    "mkdir $ANI\n",
    "cd $ANI\n",
    "\n",
    "bit-dl-ncbi-assemblies -w $ACC/pruned_acc.txt -f fasta -j 30\n",
    "# Result: ?? genomes downloaded\n",
    "gzip -d *.gz \n",
    "\n",
    "cp $PROJECT_DIR/04-analysis/NG-16087_Pseudomonas_4A2E_2000000_lib257634_5948_paired_assembly.fa $ANI/\n",
    "ls $ANI > acc_IDs.txt\n",
    "\n",
    "conda deactivate bit\n",
    "\n",
    "conda activate fastani\n",
    "#fragLen 3000\n",
    "fastANI --ql acc_IDs.txt --rl acc_IDs.txt -t 15 -k 16 --fragLen 3000 -o ANI_output_fg3000.txt --matrix\n",
    "\n",
    "awk -F '\\t' '{print $1\"\\t\"$2\"\\t\"$3}' ANI_output_fg3000.txt > ANI_output1_fg3000.txt\n",
    "awk 'BEGIN {FS=OFS=\"\\t\"} \n",
    "           {col[$1]; row[$2]; val[$2,$1]=$3}\n",
    "     END   {for(c in col) printf \"%s\", OFS c; print \"\";\n",
    "            for(r in row)\n",
    "              {printf \"%s\", r;\n",
    "               for(c in col) printf \"%s\", OFS val[r,c]\n",
    "               print \"\"}}' ANI_output1_fg3000.txt > ANI_OUT_MATRIX_fg3000.txt\n",
    "\n",
    "# using the matrix as input run the R script ANI.R\n",
    "\n",
    "##################################################################\n",
    "# Plot the heat map for ANI \n",
    "##################################################################\n",
    "# Found here `NRPS-Pnunensis4A2e/02-scripts/ANI.R`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 03- Mycin BGC - BLAST \n",
    "#### Where is the BGC also found ? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#!/bin/bash\n",
    "# running with blast4\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "\n",
    "PROJECT_DIR=NRPS-Pnunensis4A2e\n",
    "AN_DIR=$PROJECT_DIR/04-analysis/blast\n",
    "NCBI=/Net/Groups/ccdata/databases/ncbi-nt\n",
    "\n",
    "mkdir $AN_DIR\n",
    "cd $AN_DIR\n",
    "\n",
    "##################################################################\n",
    "# BLAST -remote # Downloaded on 13/04/2023\n",
    "##################################################################\n",
    "conda activate blasty\n",
    "\n",
    "cd $NCBI\n",
    "# I updated the nt database on our server by running: (Only do once)\n",
    "# And for future purposes it can be updated by running only 'update_blastdb.pl --passive --force blastdb nt'\n",
    "\n",
    "wget $NCBI \"ftp://ftp.ncbi.nlm.nih.gov/blast/db/nt.*tar.gz\"\n",
    "for f in nt.??.tar.gz; do\n",
    "tar -xvzf $f;\n",
    "done\n",
    "\n",
    "cd $AN_DIR\n",
    "\n",
    "mkdir $AN_DIR/blastn_corrected\n",
    "mkdir $AN_DIR/tblastn_corrected\n",
    "\n",
    "#blastn\n",
    "blastn -query $PROJECT_DIR/01-resources/tycCmycin_nt.fasta -db $NCBI/nt \\\n",
    "-max_target_seqs 100 \\\n",
    "-num_threads 25 \\\n",
    "-task megablast \\\n",
    "-evalue 0.05 \\\n",
    "-outfmt \"6 qseqid sseqid pident sacc mismatch sseq sstart send evalue sscinames nident mismatch length qcovs qcovhs\" \\\n",
    "-out matches_nt_blastn_fna_qcov_local_5.txt\n",
    "\n",
    "python3 /home/aibrahim/jupyter/jupyter-notebooks-uptodate/blastn_filter_hsps.py \\\n",
    "--blastn_file matches_nt_blastn_fna_qcov_local_5.txt \\\n",
    "--output_folder $AN_DIR/blastn_corrected\n",
    "\n",
    "awk -v FS='[\\t]' -v OFS='\\t' 'NR > 1 { print $2\":\"$10\"\\t\"$6 }' $AN_DIR/blastn_corrected/matches_nt_blastn_fna_qcov_local_filtered.txt > $AN_DIR/blastn_corrected/matches_combined_blastn_local.txt\n",
    "awk -F \"\\t\" '{print \">\"$1\"\\n\"$2}'  $AN_DIR/blastn_corrected/matches_combined_blastn_local.txt >  $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed 's/ /_/g' -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed 's/[][]//g' -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed 's/:/_/g' -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed 's/,//g' -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed 's/[()]//g' -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed \"s/'//g\" -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed \"s/'//g\" -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "sed \"s/;//g\" -i $AN_DIR/blastn_corrected/matches_blastn_local.fa\n",
    "\n",
    "cat $PROJECT_DIR/01-resources/tycCmycin_nt.fasta $AN_DIR/blastn_corrected/matches_blastn_local.fa > $AN_DIR/blastn_corrected/matches_blastn_local.fasta\n",
    "\n",
    "# tblastn\n",
    "tblastn -query $PROJECT_DIR/01-resources/mycin_nrps_protein.fasta -db $NCBI/nt \\\n",
    "-max_target_seqs 100 \\\n",
    "-num_threads 25 \\\n",
    "-evalue 0.05 \\\n",
    "-word_size 5 \\\n",
    "-outfmt \"6 qseqid sseqid pident sacc mismatch sseq sstart send evalue sscinames nident mismatch length qcovs qcovhsp\" \\\n",
    "-out $AN_DIR/tblastn_corrected/matches_nt_tblastn_faa_qcov_local_5.txt\n",
    "\n",
    "python3 /home/aibrahim/jupyter/jupyter-notebooks-uptodate/blastn_filter_hsps.py \\\n",
    "--blastn_file $AN_DIR/tblastn_corrected/matches_nt_tblastn_faa_qcov_local_5.txt \\\n",
    "--output_folder $AN_DIR/tblastn_corrected/\n",
    "\n",
    "awk -v FS='[\\t]' -v OFS='\\t' 'NR > 1 { print $2\":\"$10\"\\t\"$6 }' $AN_DIR/tblastn_corrected/matches_nt_blastn_fna_qcov_local_filtered.txt > $AN_DIR/tblastn_corrected/matches_combined_tblastn_local.txt\n",
    "awk -F \"\\t\" '{print \">\"$1\"\\n\"$2}'  $AN_DIR/tblastn_corrected/matches_combined_tblastn_local.txt >  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed 's/ /_/g' -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed 's/[][]//g' -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed 's/:/_/g' -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed 's/,//g' -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed 's/[()]//g' -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed \"s/'//g\" -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed \"s/'//g\" -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "sed \"s/;//g\" -i  $AN_DIR/tblastn_corrected/matches_tblastn_local.fa\n",
    "\n",
    "cat $PROJECT_DIR/01-resources/tycCmycin_aa.fasta $AN_DIR/tblastn_corrected/matches_tblastn_local.fa > $AN_DIR/tblastn_corrected/matches_tblastn_local.fasta\n",
    "\n",
    "##################################################################\n",
    "# MUSCLE\n",
    "##################################################################\n",
    "# PROBLEM of the mycin tree: The alignments turned out to be sqewed in the resulting ncbi table therfeore, the accessions and the coordinates have to be taken into consideration instead \n",
    "\n",
    "conda activate /home/aibrahim/anaconda3/envs/phylogenetics \n",
    "\n",
    "muscle -in $AN_DIR/blastn_corrected/matches_blastn_local.fasta -out $AN_DIR/blastn_corrected/matches_blastn_local.afa -maxiters 2\n",
    "muscle -in $AN_DIR/tblastn_corrected/matches_tblastn_local.fasta -out $AN_DIR/tblastn_corrected/matches_tblastn.afa -maxiters 4\n",
    "\n",
    "conda deactivate \n",
    "\n",
    "conda activate raxml\n",
    "\n",
    "# raxml hates illegal characters : remove them\n",
    "cd $AN_DIR/blastn_corrected\n",
    "raxmlHPC-PTHREADS-SSE3 -T 25 -f a -m GTRGAMMA -p 1989 -x 1989 -N 100 -s matches_blastn_local.afa -n bootstrap_matches_blastn_local.tre\n",
    "cd $AN_DIR/tblastn_corrected\n",
    "raxmlHPC-PTHREADS-SSE3 -T 25 -f a -m PROTGAMMAAUTO -p 1989 -x 1989 -N 100 -s matches_tblastn.afa -n bootstrap_matches_tblastn.tre\n",
    "\n",
    "conda deactivate \n",
    "\n",
    "##################################################################\n",
    "# Add the accession labels to the tree\n",
    "##################################################################\n",
    "\n",
    "# grab the accession ids from the blast results \n",
    "cd $AN_DIR/blastn_corrected\n",
    "awk '{print $4}' matches_nt_blastn_fna_qcov_local_filtered.txt > acc_list_blastn_local.txt\n",
    "cd $AN_DIR/tblastn_corrected\n",
    "awk '{print $4}' matches_nt_blastn_fna_qcov_local_filtered.txt > acc_list_tblastn.txt\n",
    "\n",
    "conda activate blasty\n",
    "\n",
    "# find the taxa label in NCBI\n",
    "cd $AN_DIR/blastn_corrected\n",
    "cat acc_list_blastn_local.txt | epost -db nuccore | esummary -db nuccore \\\n",
    "| xtract -pattern DocumentSummary -element Caption,Organism,Description > acc_list_blastn_local_taxid.txt\n",
    "cd $AN_DIR/tblastn_corrected\n",
    "cat acc_list_tblastn.txt | epost -db nuccore | esummary -db nuccore \\\n",
    "| xtract -pattern DocumentSummary -element Caption,Organism,Description > acc_list_tblastn_taxid.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter the tblastn results and run them on Clinker "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "\n",
    "PROJECT_DIR=NRPS-Pnunensis4A2e\n",
    "AN_DIR=$PROJECT_DIR/04-analysis/blast\n",
    "NCBI=/Net/Groups/ccdata/databases/ncbi-nt\n",
    "\n",
    "mkdir $AN_DIR\n",
    "cd $AN_DIR\n",
    "\n",
    "##################################################################\n",
    "# Filter tblastn according to sequence similarity and qcov\n",
    "##################################################################\n",
    "cd $AN_DIR/tblastn_corrected/\n",
    "\n",
    "#Filter tblastn according to sequence similarity/query coverage\n",
    "awk -v FS='[\\t]' '$13 > 98  && $3 > 75.0' matches_nt_blastn_fna_qcov_local_filtered.txt > matches_nt_blastn_fna_qcov_local_filtered_2.txt\n",
    "#Grab the accession IDs and the regions \n",
    "#Extract the sstart and send coordinates  and the tsrand orientation \n",
    "#awk -v FS='[\\t]' -v OFS='\\t' 'NR > 1 { print $2\":\"$7\"-\"$8 }' matches2.txt > matches_coords.txt\n",
    "awk -v FS='[\\t]' -v OFS='\\t' 'NR > 1 { print $2\":\"$7\"-\"$8\":\"$1}' matches_nt_blastn_fna_qcov_local_filtered_2.txt > extracted.txt\n",
    "\n",
    "# Extract the accession numbers\n",
    "#grab the acc and coordinates only from the file\n",
    "awk -F '|' '$0=$4 FS $5' extracted.txt > extracted_filt.txt\n",
    "# replace the | and : with tab\n",
    "sed -i 's/|:/\\t/g' extracted_filt.txt\n",
    "sed -i 's/:/\\t/g' extracted_filt.txt\n",
    "# replace the - with tab \n",
    "sed -i 's/-/\\t/g' extracted_filt.txt\n",
    "\n",
    "##################################################################\n",
    "# Download the gbk/gff files of the genomes \n",
    "##################################################################\n",
    "mkdir fastas\n",
    "cd fastas\n",
    "\n",
    "conda activate blasty\n",
    "\n",
    "# grab it according to coordinates \n",
    "while IFS= read -r line; do\n",
    "    # Perform a command on each line of the input file \n",
    "    id=$(echo -e \"${line}\" | cut -f 1)\n",
    "    echo \"${id}\"\n",
    "    efetch -id $id -db nuccore -format fasta > $id.fasta\n",
    "done < ../extracted_filt.txt\n",
    "\n",
    "##############################################\n",
    "# PROKKA for gene annotation\n",
    "##############################################\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate /home/aibrahim/anaconda3/envs/Prokka\n",
    "\n",
    "mkdir $AN_DIR/tblastn_corrected/prokka\n",
    "\n",
    "for F in $AN_DIR/tblastn_corrected/fastas/*.fasta; do \n",
    "  N=$(basename $F .fasta) ;\n",
    "  mkdir $AN_DIR/tblastn_corrected/prokka/$N ;\n",
    "  prokka --quiet --compliant --prefix $N --locustag $N --force --outdir $AN_DIR/tblastn_corrected/prokka/$N --cpus 28  $F ;\n",
    "done \n",
    "\n",
    "conda deactivate\n",
    "\n",
    "##############################################\n",
    "# Slice the gbk file\n",
    "##############################################\n",
    "# prepared a script that can slice gbk files by giving the gene_id and the number of surrounding genes\n",
    " \n",
    "conda activate biopython\n",
    "\n",
    "cd $AN_DIR/tblastn_corrected/prokka/\n",
    "for F in $AN_DIR/tblastn_corrected/prokka/*/*.gbk; do \n",
    "N=$(basename $F .gbk)\n",
    "cd $AN_DIR/tblastn_corrected/prokka/$N\n",
    "python3 $PROJECT_DIR/02-scripts/gbk_slicer.py $F $N.sliced3.gb tycC_3 41 42 ; done\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "##############################################\n",
    "# Clinker\n",
    "##############################################\n",
    "mkdir $AN_DIR/tblastn_corrected/clinker\n",
    "cd $AN_DIR/tblastn_corrected/clinker\n",
    "\n",
    "conda activate clinker\n",
    "\n",
    "clinker $AN_DIR/tblastn_corrected/prokka/*/*sliced3.gb -p mycin3.v2.html -j 28 -o mycin3.v2.aln\n",
    "clinker $AN_DIR/tblastn_corrected/prokka/*/*sliced3.gb -p mycin3.html -j 50 -o mycin3.aln\n",
    "clinker $AN_DIR/tblastn_corrected/prokka/CP007410.1/*.sliced2.gb $AN_DIR/tblastn_corrected/prokka/pnunsis4A2E/*.sliced.gb -p mycin2.html -j 50 -o mycin2.aln\n",
    "clinker $AN_DIR/tblastn_corrected/prokka/*/*sliced.gb -p mycin4.html -j 50 -o mycin4.aln\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
